app/models/api/withdrawal_attributes.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module Api
+  class WithdrawalAttributes
+    include ActiveModel::AttributeAssignment
+    include ActiveModel::Attributes
+    include ActiveModel::Model
+
+    include DatesHelper
+
+    validate :withdraw_date_valid
+    validates :withdraw_reasons_details, length: { maximum: 1000 }, allow_blank: true
+    validates :withdraw_reasons_dfe_details, length: { maximum: 1000 }, allow_blank: true
+
+    validates :reasons, presence: true
+    validate :unknown_exclusively
+
+    attr_accessor :trainee
+
+    ATTRIBUTES = %i[
+      reasons
+      withdraw_date
+      withdraw_reasons_details
+      withdraw_reasons_dfe_details
+    ].freeze
+
+    ATTRIBUTES.each do |attr|
+      attribute attr
+    end
+
+    def initialize(trainee:)
+      @trainee = trainee
+      super
+    end
+
+    def save!
```

Comment:
I am wondering whether this method should be extracted into a separate service class. That way `WithdrawAttributes` would be more like a plain data object.

Reply:
hmmm so you want something like this?


```ruby

results = WithdrawService.call

if results.success?
  render(json: { data: results.as_json }, status: :accepted)
else
  render_validation_errors(errors: results.errors)
end

```

Is this the bigger picture you want to aim for?

Reply:
Yes I like that better because it separates the two concerns here nicely. And it might be a more sensible pattern where there is more business logic for other endpoints. What do you reckon?

Reply:
I agree, having these attributes objects as simple data containers makes sense to me.

Reply:
In most cases, I imagine you could just delegate the save method directly to the ApplicationRecord. Then any logic around the saving (like queuing jobs etc) could be triggered with callbacks. You could still use a service to do the work, but you wouldn't have to remember to use it in these cases.